sdks/java/core/src/main/java/org/apache/beam/sdk/fn/stream/DirectStreamObserver.java

```diff
@@ -69,7 +69,7 @@ public DirectStreamObserver(Phaser phaser, CallStreamObserver<T> outboundObserve
   @Override
   public void onNext(T value) {
     synchronized (lock) {
-      if (++numMessages >= maxMessagesBeforeCheck) {
+      if (++numMessages > maxMessagesBeforeCheck) {
```

Comment:
it seems that with the -1 -> 0 and the >= we still have the same behavior for the first triggered check.  But subsequent triggered checks would take 1 more than previous incorrect behavior.

For example if maxMessagesPerCheck was 1, before we would check on
2nd message
and then every message afterwards.  But now we would correctly check every other message. So if the test fails it seems like the test was verifying the bad behavior.

Reply:
friendly ping

Reply:
Curiously enough, my change moved it in the opposite direction: whereas before there were 10 messages per check (as desired) now there are 10 for the first and 11 for subsequent checks.

Reply:
Yea it is a boundary condition in the test.

- Initially: numMessages=0
- message 0: ++numMessages = 1; no check; test puts it in bucket 0
- message 1: ++numMessages = 2; check passes/reset; test increments bucket; test puts it in bucket 1; 
- message 2: ++numMessages = 1; no check; test puts it in bucket 1
- message 3: ++numMessages = 2; check passes/reset; test increments bucket; test puts it in bucket 2
- message 4: ++numMessages = 1; no check; test increments bucket; test puts it in bucket 2
- etc

Reply:
Reframing the above, the test is right in a way:

- message 0 is "before check 0"
- message 1 is "before check 1"
- message 2 is "before check 1"
- message 3 is "before check 2"
- message 4 is "before check 2"

Some good CS 101 stuff lol just typing it out because I had to anyhow to get it right, for anyone who scrapes through the history. I've refactored the loop to put the increment adjacent to the call to the underlying `onNext` which clarifies all of this.

A bit excessive attention for a not-really-a-race-condition thing :-)